File: HOT100p2/25.k-个一组翻转链表.cpp
/*
 * @lc app=leetcode.cn id=25 lang=cpp
 *
 * [25] K 个一组翻转链表
 */
struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {};
};

// @lc code=start
#pragma once
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
#pragma GCC optimize("Ofast,no-stack-protector,unroll-loops")
#define ALL(v) v.begin(), v.end()
#define For(i, _) for (int i = 0, i##end = _; i < i##end; ++i)
#define FOR(i, _, __) for (int i = _, i##end = __; i < i##end; ++i)
#define Rep(i, _) for (int i = (_); i >= 0; --i)
#define REP(i, __, _) for (int i = (__), i##end = _; i >= i##end; --i)
typedef long long ll;
typedef unsigned long long ull;
#define V vector
#define pb push_back
#define pf push_front
#define qb pop_back
#define qf pop_front
#define eb emplace_back
typedef pair<int, int> pii;
typedef pair<ll, int> pli;
#define fi first
#define se second
const int dir[4][2] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
const int inf = 0x3f3f3f3f;
const ll infl = 0x3f3f3f3f3f3f3f3fll;
const int mod = 1e9 + 7;
template <class T>
inline bool ckmin(T &x, const T &y) { return x > y ? (x = y, true) : false; }
template <class T>
inline bool ckmax(T &x, const T &y) { return x < y ? (x = y, true) : false; }
int init = []()
{ ios::sync_with_stdio(false); cin.tie(nullptr); return 0; }();

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        int n = 0;
        auto cur = head;
        while (cur) {
            cur = cur->next;
            n ++;
        }
        ListNode dummy(0);
        dummy.next = head;
        ListNode *p0 = &dummy, *pre = nullptr;
        cur = head;
        while (n >= k) {
            FOR(i, 0, k) {
                auto tmp = cur->next;
                cur->next = pre;
                pre = cur;
                cur = tmp;
            }
            auto nxt = p0->next;
            p0->next->next = cur;
            p0->next = pre;
            p0 = nxt;
            n -= k;
        }
        return dummy.next;
    }
};
// @lc code=end

File: 面经150/面经150p1/8.链表/25.k-个一组翻转链表.py
#
# @lc app=leetcode.cn id=25 lang=python3
#
# [25] K 个一组翻转链表
#
from mytools import *
# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = cur = ListNode(next=head)
        def rev_(ppre: Optional[ListNode], p: Optional[ListNode]):
            cnt = k
            pre = None
            cur = p
            while cnt:
                nxt = cur.next
                cur.next = pre
                pre = cur
                cur = nxt
                cnt -= 1
            ppre.next.next = cur
            tmp = ppre.next
            ppre.next = pre
            return tmp, cur
        sz = 0
        p = head
        while p: sz += 1; p = p.next
        t = int(sz / k)
        p = head
        while t: cur, p = rev_(cur, p); t -= 1
        return dummy.next
# @lc code=end

